File: showcase/starters/langgraph-typescript/agent/frontend-tools.ts
/**
 * LangGraph TypeScript agent backing the Frontend Tools (In-App Actions) demo.
 *
 * The demo is about frontend tools — the agent has no custom backend tools.
 * CopilotKit forwards the frontend tool schemas to the agent at runtime via
 * `state.copilotkit.actions`; the agent binds them when invoking the model,
 * and the handler executes in the browser.
 */

import { RunnableConfig } from "@langchain/core/runnables";
import { SystemMessage } from "@langchain/core/messages";
import { MemorySaver, START, StateGraph } from "@langchain/langgraph";
import { ChatOpenAI } from "@langchain/openai";
import {
  convertActionsToDynamicStructuredTools,
  CopilotKitStateAnnotation,
} from "@copilotkit/sdk-js/langgraph";

const AgentStateAnnotation = CopilotKitStateAnnotation;
export type AgentState = typeof AgentStateAnnotation.State;

const SYSTEM_PROMPT = "You are a helpful, concise assistant.";

async function chatNode(state: AgentState, config: RunnableConfig) {
  const model = new ChatOpenAI({ temperature: 0, model: "gpt-4o-mini" });

  const modelWithTools = model.bindTools!([
    ...convertActionsToDynamicStructuredTools(state.copilotkit?.actions ?? []),
  ]);

  const response = await modelWithTools.invoke(
    [new SystemMessage({ content: SYSTEM_PROMPT }), ...state.messages],
    config,
  );

  return { messages: response };
}

const workflow = new StateGraph(AgentStateAnnotation)
  .addNode("chat_node", chatNode)
  .addEdge(START, "chat_node")
  .addEdge("chat_node", "__end__");

const memory = new MemorySaver();

export const graph = workflow.compile({
  checkpointer: memory,
});

File: showcase/integrations/langgraph-fastapi/src/agents/src/frontend_tools.py
"""LangGraph agent backing the Chat Customization (CSS) demo.

The demo is about CSS — the agent has no custom tools or behavior.
CopilotKitMiddleware is attached so CopilotKit-specific context is
picked up if the frontend ever registers suggestions/components.
"""

from langchain.agents import create_agent
from langchain_openai import ChatOpenAI
from copilotkit import CopilotKitMiddleware

graph = create_agent(
    model=ChatOpenAI(model="gpt-4o-mini"),
    tools=[],
    middleware=[CopilotKitMiddleware()],
    system_prompt="You are a helpful, concise assistant.",
)
